src/freshrssapi.cpp

```diff
@@ -0,0 +1,546 @@
+#include "freshrssapi.h"
+
+#include <cstring>
+#include <curl/curl.h>
+#include <json.h>
+#include <time.h>
+#include <vector>
+
+#include "config.h"
+#include "strprintf.h"
+#include "utils.h"
+#include "rss/feed.h"
+#include "3rd-party/json.hpp"
+
+#define FRESHRSS_LOGIN "/accounts/ClientLogin"
+#define FRESHRSS_API_PREFIX "/reader/api/0/"
+#define FRESHRSS_FEED_PREFIX "/reader/api/0/stream/contents/"
+
+#define FRESHRSS_OUTPUT_SUFFIX "?output=json"
+
+#define FRESHRSS_SUBSCRIPTION_LIST \
+	FRESHRSS_API_PREFIX "subscription/list" FRESHRSS_OUTPUT_SUFFIX
+#define FRESHRSS_API_MARK_ALL_READ_URL FRESHRSS_API_PREFIX "mark-all-as-read"
+#define FRESHRSS_API_EDIT_TAG_URL FRESHRSS_API_PREFIX "edit-tag"
+#define FRESHRSS_API_TOKEN_URL FRESHRSS_API_PREFIX "token"
+
+namespace newsboat {
+
+FreshRssApi::FreshRssApi(ConfigContainer* c)
+	: RemoteApi(c)
+{
+	/* server = cfg->get_configvalue("freshrss-url"); */
+}
+
+bool FreshRssApi::authenticate()
+{
+	auth = retrieve_auth();
+	LOG(Level::DEBUG, "FreshRssApi::authenticate: Auth = %s", auth);
+	return auth != "";
+}
+
+static size_t my_write_data(void* buffer, size_t size, size_t nmemb,
+	void* userp)
+{
+	std::string* pbuf = static_cast<std::string*>(userp);
+	pbuf->append(static_cast<const char*>(buffer), size * nmemb);
+	return size * nmemb;
+}
+
+std::string FreshRssApi::retrieve_auth()
+{
+	CURL* handle = curl_easy_init();
+	Credentials cred = get_credentials("freshrss", "FreshRSS");
+	if (cred.user.empty() || cred.pass.empty()) {
+		return "";
+	}
+
+	char* username = curl_easy_escape(handle, cred.user.c_str(), 0);
+	char* password = curl_easy_escape(handle, cred.pass.c_str(), 0);
+
+	std::string postcontent = strprintf::fmt(
+			"service=reader&Email=%s&Passwd=%s&source=%s%2F%s&accountType="
+			"HOSTED_OR_GOOGLE&continue=http://www.google.com/",
```

Comment:
Suggestion: does FreshRSS actually need all those parameters? I realize you copied them from TheOldReader, and they work, but there's no reason to send more stuff than the service requires.

Reply:
Mostly not needed, at least for FreshRSS